### PR TITLE
Feature: dependency checks based on diagrams for nx and lerna monorepos

### PR DESCRIPTION
**Pull request by torfmaster**
# Content of the Pull Request

This pull requests features dependency checks between packages of nx/lerna monorepos using PlantUML diagrams.

# Implementation

The feature extends the graph api by supplying an alternative source for dependency graphs.

# Tests

The feature is integration tested.

### CODE REVIEW

**Comment by torfmaster** [comment]
@c-harding would you volunteer for a review? I think you the only maintainer of the project.

File: src/common/extraction/extractNxGraph.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import * as fs from "fs"
+import * as path from "path"
+import { Graph } from "./graph"
+
+type Nodes = {
+	[index: string]: [Outgoing]
```
Comment (c-harding) [inline src/common/extraction/extractNxGraph.ts:6]: What's the point of using singleton arrays here? Why not `[index: string]: Outgoing`?

In other words, why is this type `Record<string, [Outgoing]>` and not `Record<string, Outgoing>`?

File: src/common/extraction/extractNxGraph.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+type Outgoing = {
+	source: string
+	target: string
+}
+
+export function extractNxGraph(rootFolder?: string): Graph {
+	const workspaceRoot = rootFolder ?? path.dirname(guessLocationOfNxConfig()!)
```
Comment (c-harding) [inline src/common/extraction/extractNxGraph.ts:15]: Would it be better to throw a human-readable error if `guessLocationOfNxConfig()` is undefined? At the moment, dirname will throw a TypeError

File: README.md [hunk 3]
```diff
@@ -92,3 +92,32 @@ The path of the project is always relative to a given `tsconfig.json`.
...
+
+describe("architecture", ()=> {
+    jest.setTimeout(60000);
+
+    it('the architecture adheres to the diagram', async () => {
+    	const diagramLocation = path.resolve('docs', 'components.puml');
+    
+    	const rule = await slicesOfNxProject()
+            .asTheyAre(true)
+    		.should()
```
Comment (c-harding) [inline README.md:116]: This API is not especially clear, particularly given that `asTheyAre(true)` and `identity(true)` are not true identity/homomorphic functions, as they perform filtering. I would instead suggest `slicesOfNxProject().withoutExternals().should()` in this API, and `SliceConditionBuilder` could extend `SlicesShouldCondition` so that, when externals are not filtered out, `slicesOfNxProject().should()` can be used directly.

Please then also replace `identity(true)` with `removeExternals`, and `identity(false)` with `identity`.

File: test/architecture.spec.ts
Comment (c-harding) [inline test/architecture.spec.ts:1]: Why did you delete this test?

**Review by c-harding — COMMENTED** [review]
Here are my first comments from looking through the code; I'll give it a more thorough test next week.

File: test/architecture.spec.ts
Comment (torfmaster) [reply test/architecture.spec.ts:1]: That was a mistake, I restored it.

File: src/common/extraction/extractNxGraph.ts (hunk 2 above)
Comment (torfmaster) [reply src/common/extraction/extractNxGraph.ts]: Good catch! I fixed that.

File: src/common/extraction/extractNxGraph.ts (hunk 1 above)
Comment (torfmaster) [reply src/common/extraction/extractNxGraph.ts]: That was a bad habit from Haskell, it should be `Outgoing[]`.

File: README.md (hunk 3 above)
Comment (torfmaster) [reply README.md:116]: I also think that the identity should be an isomorphism. I am not sure what the value of "if I choose `withoutExternal once I don't want to see it again in the auto completion" is. I thus removed `SlicesShouldCondition` completely. I can restore it if we find this useful.

Comment (torfmaster) [reply README.md]: Thanks, I fixed that.

**Review by c-harding — APPROVED** [review]

**Comment by torfmaster** [comment]
Thank you for the review and for spotting the mistakes. I am missing the permissions to merge. Can you please merge this PR (and maybe the other approved PR as well). Thank you!

**Comment by c-harding** [comment]
Merged, version 5.4.0
